File: py/83.remove-duplicates-from-sorted-list.py
#
# @lc app=leetcode id=83 lang=python3
#
# [83] Remove Duplicates from Sorted List
#
# https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/
#
# algorithms
# Easy (42.26%)
# Total Accepted:    317.2K
# Total Submissions: 750.1K
# Testcase Example:  '[1,1,2]'
#
# Given a sorted linked list, delete all duplicates such that each element
# appear only once.
# 
# Example 1:
# 
# 
# Input: 1->1->2
# Output: 1->2
# 
# 
# Example 2:
# 
# 
# Input: 1->1->2->3->3
# Output: 1->2->3
# 
# 
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from list_node import ListNode


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if head is None:
            return None

        node = head
        while node.next is not None:
            if node.val != node.next.val:
                node = node.next
                continue

            node.next = node.next.next

        return head


if __name__ == '__main__':
    tests = [
        (ListNode(1, 2, 2, 3), ListNode(1, 2, 3)),
        (ListNode(2, 2, 2, 3), ListNode(2, 3)),
        (ListNode(2, 2, 2, 2), ListNode(2)),
    ]
    for tc in tests:
        actual = Solution().deleteDuplicates(tc[0])
        expected = tc[1]
        if actual != expected:
            print(f"deleteDuplicates({tc[0]}) == {actual} != {expected}")

File: go/83.remove-duplicates-from-sorted-list.go
package main

/*
 * @lc app=leetcode id=83 lang=golang
 *
 * [83] Remove Duplicates from Sorted List
 *
 * https://leetcode.com/problems/remove-duplicates-from-sorted-list/description/
 *
 * algorithms
 * Easy (42.00%)
 * Total Accepted:    306.3K
 * Total Submissions: 729.2K
 * Testcase Example:  '[1,1,2]'
 *
 * Given a sorted linked list, delete all duplicates such that each element
 * appear only once.
 *
 * Example 1:
 *
 *
 * Input: 1->1->2
 * Output: 1->2
 *
 *
 * Example 2:
 *
 *
 * Input: 1->1->2->3->3
 * Output: 1->2->3
 *
 *
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates_83(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	node := head.Next
	prev := head
	for node != nil {
		if node.Val == prev.Val {
			node = node.Next
			prev.Next = node
			continue
		}

		prev = node
		node = node.Next
	}
	return head
}
